lib/pdn/pdn.c

```diff
@@ -0,0 +1,457 @@
+#include <stdlib.h>
+#include <stdbool.h>
+#include <zephyr.h>
+#include <init.h>
+#include <modem/pdn.h>
+#include <modem/at_cmd.h>
+#include <modem/at_notif.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(pdn, CONFIG_PDN_LOG_LEVEL);
+
+/* Buffer size for AT command and responses.
+ * Longest AT command is:
+ *   AT+CGDCONT=20,"IPV4V6","<APN>",0,0,0,0,0,0,0,0,0
+ * where the APN can be up to 100 chars.
+ * That's 43 chars, plus 100, which we round to 160.
+ */
+#define AT_BUF_SIZE (160)
+/* Global buffer for AT commands and responses. */
+static char at_buf[AT_BUF_SIZE];
+
+#define CID_UNASSIGNED (-1)
+
+struct pdn {
+	pdn_event_handler_t callback;
+	int8_t context_id;
+};
+
+static int esm_from_notif;
+static struct k_sem notif_sem;
+static struct pdn pdn_contexts[CONFIG_PDN_CONTEXTS_MAX];
+static pdn_event_handler_t default_callback;
+
+static struct pdn *pdn_find(int cid)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pdn_contexts); i++) {
+		struct pdn *pdn = &pdn_contexts[i];
+
+		if (pdn->context_id == cid) {
+			return pdn;
+		}
+	}
+
+	return NULL;
+}
+
+static struct pdn *pdn_avail_get(void)
+{
+	return pdn_find(CID_UNASSIGNED);
+}
+
+static void at_notif_handler(void *context, const char *notif)
+{
+	char *p;
+	uint8_t cid;
+	uint8_t esm_err;
+	struct pdn *pdn;
+
+	/* These notifications are at least 10 characters long,
+	 * skip shorter strings.
+	 */
+	if (strlen(notif) < 10) {
+		return;
+	}
+
+	/* AT notif: +CNEC_ESM: <cause>,<cid>
+	 * Handle this by itself, we need to extract both CID and ESM
+	 * TODO: check before all other notifs?
+	 */
+	p = strstr(notif, "CNEC_ESM: ");
+	if (p) {
+		esm_err = atoi(p + strlen("CNEC_ESM: "));
+
+		p = strchr(p, ',');
+		if (!p) {
+			return;
+		}
+
+		cid = atoi(p + 1);
+
+		pdn = pdn_find(cid);
+		if (!pdn) {
+			return;
+		}
+
+		if (pdn->callback) {
+			pdn->callback((intptr_t)cid, PDN_EVENT_CNEC_ESM,
+				      esm_err);
+		}
+
+		esm_from_notif = esm_err;
+		k_sem_give(&notif_sem);
+
+		return;
+	}
+
+	const struct {
+		const char *notif;
+		const enum pdn_event event;
+	} map[] = {
+		{"CGEV: ME PDN ACT",	PDN_EVENT_ACTIVATED},	/* +CGEV: ME PDN ACT <cid> */
+		{"CGEV: ME PDN DEACT",	PDN_EVENT_DEACTIVATED},	/* +CGEV: ME PDN DEACT <cid> */
+		/* Order is important */
+		{"CGEV: IPV6 FAIL",	PDN_EVENT_IPV6_DOWN},	/* +CGEV: IPV6 FAIL <cid> */
+		{"CGEV: IPV6",		PDN_EVENT_IPV6_UP},	/* +CGEV: IPV6 <cid> */
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(map); i++) {
+		p = strstr(notif, map[i].notif);
+		if (!p) {
+			continue;
+		}
+
+		cid = atoi(p + strlen(map[i].notif));
+
+		if (default_callback && cid == 0) {
+			default_callback(0, map[i].event, 0);
+			return;
+		}
+
+		pdn = pdn_find(cid);
+		if (!pdn) {
+			return;
+		}
+
+		if (pdn->callback) {
+			pdn->callback((intptr_t)cid, map[i].event, 0);
+			return;
+		}
+	}
+}
+
+int pdn_init(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pdn_contexts); i++) {
+		pdn_contexts[i].context_id = CID_UNASSIGNED;
+	}
+
+	k_sem_init(&notif_sem, 0, 1);
```

Comment:
Should there be some sort of protection here agains re-running `pdn_init()` if it's already initialized? Can imagine it will cause trouble if the context list is overeritten and semaphore is re-initialized to zero when it's actually 1

Reply:
I have added this